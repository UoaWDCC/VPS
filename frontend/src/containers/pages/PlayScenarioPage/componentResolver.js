/* eslint-disable no-param-reassign */
import React from "react";

// Shared components from authoring canvas
import ButtonComponent from "../AuthoringTool/Components/ButtonComponent";
import TextComponent from "../AuthoringTool/Components/TextComponent";
import ImageComponent from "../AuthoringTool/Components/ImageComponent";
import FirebaseImageComponent from "../AuthoringTool/Components/FirebaseImageComponent";

// Components for playing only
import AudioComponent from "./Components/AudioComponent";

/**
 * This function returns the appropriate React component for a scene component object when playing
 *
 * @example
 * {currentScene?.components?.map((component, index) =>
 *   componentResolver(component, index, selectElement)
 * )}
 */
export default function componentResolver(component, index, onClick) {
  switch (component.type) {
    // ADD NEW COMPONENT TYPES HERE
    case "BUTTON":
      return (
        <ButtonComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "TEXT":
      return (
        <TextComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "IMAGE":
      return (
        <ImageComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "FIREBASEIMAGE":
      return (
        <FirebaseImageComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "FIREBASEAUDIO":
      return (
        <AudioComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    default:
      break;
  }

  return <div />;
}
