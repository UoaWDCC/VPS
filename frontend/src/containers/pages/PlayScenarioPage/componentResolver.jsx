/* eslint-disable no-param-reassign */

// Shared components from authoring canvas
import ButtonComponent from "../AuthoringTool/Components/ButtonComponent";
import FirebaseImageComponent from "../AuthoringTool/Components/FirebaseImageComponent";
import ImageComponent from "../AuthoringTool/Components/ImageComponent";
import TextComponent from "../AuthoringTool/Components/TextComponent";

// Components for playing only
import AudioComponent from "./Components/AudioComponent";

/**
 * This function returns the appropriate React component for a scene component object when playing
 *
 * @example
 * {currentScene?.components?.map((component, index) =>
 *   componentResolver(component, index, selectElement)
 * )}
 */
export default function componentResolver(component, index, onClick) {
  switch (component.type) {
    // ADD NEW COMPONENT TYPES HERE
    case "BUTTON":
      return (
        <ButtonComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "SPEECHTEXT":
    case "TEXT":
      return (
        <TextComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "IMAGE":
      return (
        <ImageComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "FIREBASEIMAGE":
      return (
        <FirebaseImageComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    case "FIREBASEAUDIO":
      return (
        <AudioComponent
          key={component.id}
          id={index}
          onClick={onClick}
          component={component}
        />
      );
    default:
      break;
  }

  return <div />;
}
