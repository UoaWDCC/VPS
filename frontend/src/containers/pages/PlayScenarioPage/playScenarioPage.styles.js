import { makeStyles } from "@material-ui/core";
/**
 * This function contains the styles used in the PlayScenarioPage
 */
export default function useStyles() {
  return makeStyles({
    canvasContainer: {
      height: "100vh",
      width: "100vw",
    },
  })();
}
