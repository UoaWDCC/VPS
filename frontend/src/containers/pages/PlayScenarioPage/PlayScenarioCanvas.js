import React, { useState, useContext } from "react";
import { useGet, usePost, usePut } from "../../../hooks/crudHooks";
import componentResolver from "./componentResolver";
import PlayScenarioContext from "../../../context/PlayScenarioContext";
import ProgressBar from "./progressBar";
import CountdownTimer from "../../../components/TimerComponent";
import AuthenticationContext from "../../../context/AuthenticationContext";

/**
 * This component displays the scene components on the screen when playing a scenario
 *
 * @component
 */
export default function PlayScenarioCanvas(props) {
  const { progress, graph } = props;
  const [currentScene, setCurrentScene] = useState(null);
  const [maxProgress, setMaxProgress] = useState(0);
  // eslint-disable-next-line no-unused-vars
  const { scenarioId, currentSceneId, setCurrentSceneId } =
    useContext(PlayScenarioContext);

  const { user } = useContext(AuthenticationContext);

  useGet(
    `/api/scenario/${scenarioId}/scene/full/${currentSceneId}`,
    setCurrentScene,
    false
  );

  const componentOnClick = (component) => {
    if (component.type === "BUTTON" && component.nextScene !== "") {
      setMaxProgress(Math.max(progress, maxProgress));
      graph.visit(component.nextScene);
      if (graph.isEndScene(component.nextScene)) {
        // update db

        const userData = {
          name: user.displayName,
          uid: user.uid,
          email: user.email,
        };

        usePost("/api/user", userData);
      }
      setCurrentSceneId(component.nextScene);
    }
  };

  return (
    <>
      {currentScene?.components?.map((component, index) =>
        componentResolver(component, index, () => componentOnClick(component))
      )}
      <ProgressBar value={Math.max(progress, maxProgress) * 100} />
      {currentScene?.time ? (
        <CountdownTimer
          targetDate={new Date().setSeconds(
            new Date().getSeconds() + currentScene?.time
          )}
        />
      ) : null}
    </>
  );
}
