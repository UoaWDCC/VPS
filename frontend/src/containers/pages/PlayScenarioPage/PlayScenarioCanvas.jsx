import CountdownTimer from "../../../components/TimerComponent";
import componentResolver from "./componentResolver";

/**
 * This component displays the scene components on the screen when playing a scenario
 *
 * @component
 */

export default function PlayScenarioCanvas({
  scene,
  incrementor,
  reset,
  setAddFlags,
  setRemoveFlags,
}) {
  return (
    <>
      {scene.components?.map((component, index) => {
        let action = () =>
          component.nextScene && incrementor(component.nextScene);
        switch (component.type) {
          case "RESET_BUTTON":
            action = reset;
            break;
          case "BUTTON":
            setAddFlags(component.flagAdditions);
            setRemoveFlags(component.flagDeletions);
            break;
          default:
            break;
        }

        return componentResolver(component, index, action);
      })}
      {scene.time && (
        <CountdownTimer
          targetDate={new Date().setSeconds(
            new Date().getSeconds() + scene.time
          )}
          sceneTime={scene.time}
        />
      )}
    </>
  );
}
