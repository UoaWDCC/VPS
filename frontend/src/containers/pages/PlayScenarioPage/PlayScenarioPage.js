import React, { useContext } from "react";
import useStyles from "./playScenarioPage.styles";
import PlayScenarioContext from "../../../context/PlayScenarioContext";
import PlayScenarioCanvas from "./PlayScenarioCanvas";
import LoadingPage from "../LoadingPage";
import ScenarioPreloader from "./Components/ScenarioPreloader";

export default function PlayScenarioPage() {
  const styles = useStyles();
  const { currentSceneId } = useContext(PlayScenarioContext);

  if (currentSceneId === null) {
    return <LoadingPage text="Loading contents..." />;
  }

  return (
    <>
      {currentSceneId && (
        <div className={styles.canvasContainer}>
          <div className={styles.canvas}>
            <PlayScenarioCanvas />
          </div>
        </div>
      )}
      <ScenarioPreloader />
    </>
  );
}
