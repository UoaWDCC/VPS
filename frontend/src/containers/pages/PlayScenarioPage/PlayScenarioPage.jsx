import { useContext } from "react";
import PlayScenarioContext from "../../../context/PlayScenarioContext";
import useGraph from "../../../hooks/useGraph";
import LoadingPage from "../LoadingPage";
import ScenarioPreloader from "./Components/ScenarioPreloader";
import PlayScenarioCanvas from "./PlayScenarioCanvas";
import useStyles from "./playScenarioPage.styles";

/**
 * This page allows users to play a scenario.
 *
 * @container
 */
export default function PlayScenarioPage() {
  const styles = useStyles();
  const { currentSceneId, scenarioId } = useContext(PlayScenarioContext);
  const { isLoading, graph } = useGraph(scenarioId);

  if (currentSceneId === null || isLoading) {
    return <LoadingPage text="Loading contents..." />;
  }

  return (
    <>
      {currentSceneId && (
        <div className={styles.canvasContainer}>
          <div className={styles.canvas}>
            <PlayScenarioCanvas
              progress={graph.progress(currentSceneId)}
              graph={graph}
            />
          </div>
        </div>
      )}
      {window.location === window.parent.location ? (
        <ScenarioPreloader key={1} />
      ) : null}
    </>
  );
}
