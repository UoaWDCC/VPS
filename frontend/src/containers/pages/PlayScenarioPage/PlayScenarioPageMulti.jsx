import { useContext } from "react";
import { useParams, useHistory } from "react-router-dom";
import AuthenticationContext from "context/AuthenticationContext";
import { usePost, usePut } from "hooks/crudHooks";
import LoadingPage from "../LoadingPage";
import NotesDisplayCard from "../../../components/NotesDisplayCard";
import ScenarioPreloader from "./Components/ScenarioPreloader";
import PlayScenarioCanvas from "./PlayScenarioCanvas";
import useStyles from "./playScenarioPage.styles";

/**
 * This page allows users to play a multiplayer scenario.
 *
 * @container
 */
export default function PlayScenarioPageMulti({ graph, group }) {
  const { user, getUserIdToken: token } = useContext(AuthenticationContext);
  const { scenarioId, sceneId } = useParams();
  const history = useHistory();
  const styles = useStyles();

  const currScene = graph?.getScene(sceneId);

  if (!currScene || !group) return <LoadingPage text="Loading contents..." />;

  const incrementor = (nextSceneId) => {
    graph.visit(nextSceneId);
    if (graph.isEndScene(nextSceneId)) {
      const path = graph.getPath();
      usePut(`/api/user/${user.uid}`, { scenarioId, path }, token);
      path.forEach((id) => {
        usePut(`/api/scenario/${scenarioId}/scene/visited/${id}`, {}, token);
      });
    }
    history.replace(`/play/${scenarioId}/multiplayer/${nextSceneId}`);
  };

  const validatedIncrementor = async (nextSceneId) => {
    const res = await usePost(
      `/api/group/path/${group._id}`,
      { currentSceneId: sceneId, nextSceneId },
      token
    );
    if (res === "Scene added to path") incrementor(nextSceneId);
    else history.replace(`/play/desync`);
  };

  return (
    <>
      <div className={styles.canvasContainer}>
        <div className={styles.canvas}>
          <PlayScenarioCanvas
            progress={graph.progress(sceneId)}
            scene={currScene}
            incrementor={validatedIncrementor}
          />
        </div>
      </div>
      {window.location === window.parent.location && (
        <ScenarioPreloader scenarioId={scenarioId} graph={graph} key={1} />
      )}{" "}
      <NotesDisplayCard group={group} user={user} />
    </>
  );
}
