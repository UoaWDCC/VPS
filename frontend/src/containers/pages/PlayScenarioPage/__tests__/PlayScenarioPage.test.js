import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import PlayScenarioPage from "../PlayScenarioPage";
import PlayScenarioContext from "../../../../context/PlayScenarioContext";

test("PlayScenario page snapshot test", () => {
  const context = {
    currentSceneId: "EXISTS",
  };

  const { baseElement } = render(
    <BrowserRouter>
      <PlayScenarioContext.Provider value={context}>
        <PlayScenarioPage />
      </PlayScenarioContext.Provider>
    </BrowserRouter>
  );

  expect(baseElement).toMatchSnapshot();
});

test("PlayScenario page when no current SceneID snapshot test", () => {
  const context = {
    currentSceneId: null,
  };

  const { baseElement } = render(
    <BrowserRouter>
      <PlayScenarioContext.Provider value={context}>
        <PlayScenarioPage />
      </PlayScenarioContext.Provider>
    </BrowserRouter>
  );

  expect(baseElement).toMatchSnapshot();
});
