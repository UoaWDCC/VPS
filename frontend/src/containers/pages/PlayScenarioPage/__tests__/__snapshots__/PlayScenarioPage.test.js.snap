// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayScenario page snapshot test 1`] = `
<body>
  <div>
    <div
      class="makeStyles-canvasContainer-1"
    >
      <div />
    </div>
  </div>
</body>
`;

exports[`PlayScenario page when no current SceneID snapshot test 1`] = `
<body>
  <div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center"
      style="min-height: 100vh;"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-center"
      >
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <div
            class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
            role="progressbar"
            style="width: 40px; height: 40px; color: rgb(0, 138, 123);"
          >
            <svg
              class="MuiCircularProgress-svg"
              viewBox="22 22 44 44"
            >
              <circle
                class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item"
        >
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            Loading contents...
          </p>
        </div>
      </div>
    </div>
  </div>
</body>
`;
